File: python/tests/test_mask_pii.py
import os
import sys
import unittest

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "src"))

from mask_pii import Masker


class MaskPiiTestCase(unittest.TestCase):
    def assert_cases(self, masker, cases):
        for input_text, expected in cases:
            with self.subTest(input=input_text):
                self.assertEqual(masker.process(input_text), expected)

    def test_email_basic_cases(self):
        masker = Masker().mask_emails()
        self.assert_cases(
            masker,
            [
                ("alice@example.com", "a****@example.com"),
                ("a@b.com", "*@b.com"),
                ("ab@example.com", "a*@example.com"),
                ("a.b+c_d@example.co.jp", "a******@example.co.jp"),
            ],
        )

    def test_email_mixed_text(self):
        masker = Masker().mask_emails()
        self.assert_cases(
            masker,
            [
                ("Contact: alice@example.com.", "Contact: a****@example.com."),
                (
                    "alice@example.com and bob@example.org",
                    "a****@example.com and b**@example.org",
                ),
            ],
        )

    def test_email_edge_cases(self):
        masker = Masker().mask_emails()
        self.assert_cases(
            masker,
            [
                ("alice@example", "alice@example"),
                ("alice@localhost", "alice@localhost"),
                ("alice@@example.com", "alice@@example.com"),
                (
                    "first.last+tag@sub.domain.com",
                    "f*************@sub.domain.com",
                ),
            ],
        )

    def test_phone_basic_formats(self):
        masker = Masker().mask_phones()
        self.assert_cases(
            masker,
            [
                ("090-1234-5678", "***-****-5678"),
                ("Call (555) 123-4567", "Call (***) ***-4567"),
                ("Intl: +81 3 1234 5678", "Intl: +** * **** 5678"),
                ("+1 (800) 123-4567", "+* (***) ***-4567"),
            ],
        )

    def test_phone_short_and_boundary_lengths(self):
        masker = Masker().mask_phones()
        self.assert_cases(
            masker,
            [
                ("1234", "1234"),
                ("12345", "*2345"),
                ("12-3456", "**-3456"),
            ],
        )

    def test_phone_mixed_text(self):
        masker = Masker().mask_phones()
        self.assert_cases(
            masker,
            [
                (
                    "Tel: 090-1234-5678 ext. 99",
                    "Tel: ***-****-5678 ext. 99",
                ),
                (
                    "Numbers: 111-2222 and 333-4444",
                    "Numbers: ***-2222 and ***-4444",
                ),
            ],
        )

    def test_phone_edge_cases(self):
        masker = Masker().mask_phones()
        self.assert_cases(
            masker,
            [
                ("abcdef", "abcdef"),
                ("+", "+"),
                ("(12) 345 678", "(**) **5 678"),
            ],
        )

    def test_combined_masking(self):
        masker = Masker().mask_emails().mask_phones()
        self.assert_cases(
            masker,
            [
                (
                    "Contact: alice@example.com or 090-1234-5678.",
                    "Contact: a****@example.com or ***-****-5678.",
                ),
                (
                    "Email bob@example.org, phone +1 (800) 123-4567",
                    "Email b**@example.org, phone +* (***) ***-4567",
                ),
            ],
        )

    def test_custom_mask_character(self):
        email_masker = Masker().mask_emails().with_mask_char("#")
        phone_masker = Masker().mask_phones().with_mask_char("#")
        combined = Masker().mask_emails().mask_phones().with_mask_char("#")

        self.assert_cases(
            email_masker,
            [("alice@example.com", "a####@example.com")],
        )
        self.assert_cases(
            phone_masker,
            [("090-1234-5678", "###-####-5678")],
        )

        got = combined.process("Contact: alice@example.com or 090-1234-5678.")
        want = "Contact: a####@example.com or ###-####-5678."
        self.assertEqual(got, want)

    def test_masker_configuration(self):
        input_text = "alice@example.com 090-1234-5678"

        passthrough = Masker()
        self.assertEqual(passthrough.process(input_text), input_text)

        email_only = Masker().mask_emails()
        self.assertEqual(
            email_only.process(input_text),
            "a****@example.com 090-1234-5678",
        )

        phone_only = Masker().mask_phones()
        self.assertEqual(
            phone_only.process(input_text),
            "alice@example.com ***-****-5678",
        )

        both = Masker().mask_emails().mask_phones()
        self.assertEqual(
            both.process(input_text),
            "a****@example.com ***-****-5678",
        )

    def test_non_ascii_text_is_preserved(self):
        masker = Masker().mask_emails().mask_phones()
        input_text = "連絡先: alice@example.com と 090-1234-5678"
        expected = "連絡先: a****@example.com と ***-****-5678"
        self.assertEqual(masker.process(input_text), expected)


if __name__ == "__main__":
    unittest.main()

File: lua/tests/test_mask_pii.lua
local function script_dir()
  local info = debug.getinfo(1, "S")
  local source = info.source
  if source:sub(1, 1) == "@" then
    source = source:sub(2)
  end
  return source:match("(.*/)") or "./"
end

local root = script_dir() .. "../"
package.path = root .. "src/?.lua;" .. root .. "src/?/init.lua;" .. package.path

local mask_pii = require("mask_pii")
local Masker = mask_pii.Masker

local function assert_equal(actual, expected, message)
  if actual ~= expected then
    local prefix = message or "assertion failed"
    error(prefix .. ": expected " .. tostring(expected) .. ", got " .. tostring(actual))
  end
end

local function assert_cases(masker, cases)
  for _, case in ipairs(cases) do
    local input_text = case[1]
    local expected = case[2]
    local got = masker:process(input_text)
    assert_equal(got, expected, "case failed for input: " .. input_text)
  end
end

local function run()
  local masker

  masker = Masker.new():mask_emails()
  assert_cases(masker, {
    { "alice@example.com", "a****@example.com" },
    { "a@b.com", "*@b.com" },
    { "ab@example.com", "a*@example.com" },
    { "a.b+c_d@example.co.jp", "a******@example.co.jp" },
  })

  masker = Masker.new():mask_emails()
  assert_cases(masker, {
    { "Contact: alice@example.com.", "Contact: a****@example.com." },
    { "alice@example.com and bob@example.org", "a****@example.com and b**@example.org" },
  })

  masker = Masker.new():mask_emails()
  assert_cases(masker, {
    { "alice@example", "alice@example" },
    { "alice@localhost", "alice@localhost" },
    { "alice@@example.com", "alice@@example.com" },
    { "first.last+tag@sub.domain.com", "f*************@sub.domain.com" },
  })

  masker = Masker.new():mask_phones()
  assert_cases(masker, {
    { "090-1234-5678", "***-****-5678" },
    { "Call (555) 123-4567", "Call (***) ***-4567" },
    { "Intl: +81 3 1234 5678", "Intl: +** * **** 5678" },
    { "+1 (800) 123-4567", "+* (***) ***-4567" },
  })

  masker = Masker.new():mask_phones()
  assert_cases(masker, {
    { "1234", "1234" },
    { "12345", "*2345" },
    { "12-3456", "**-3456" },
  })

  masker = Masker.new():mask_phones()
  assert_cases(masker, {
    { "Tel: 090-1234-5678 ext. 99", "Tel: ***-****-5678 ext. 99" },
    { "Numbers: 111-2222 and 333-4444", "Numbers: ***-2222 and ***-4444" },
  })

  masker = Masker.new():mask_phones()
  assert_cases(masker, {
    { "abcdef", "abcdef" },
    { "+", "+" },
    { "(12) 345 678", "(**) **5 678" },
  })

  masker = Masker.new():mask_emails():mask_phones()
  assert_cases(masker, {
    { "Contact: alice@example.com or 090-1234-5678.", "Contact: a****@example.com or ***-****-5678." },
    { "Email bob@example.org, phone +1 (800) 123-4567", "Email b**@example.org, phone +* (***) ***-4567" },
  })

  local email_masker = Masker.new():mask_emails():with_mask_char("#")
  local phone_masker = Masker.new():mask_phones():with_mask_char("#")
  local combined = Masker.new():mask_emails():mask_phones():with_mask_char("#")

  assert_cases(email_masker, { { "alice@example.com", "a####@example.com" } })
  assert_cases(phone_masker, { { "090-1234-5678", "###-####-5678" } })

  assert_equal(
    combined:process("Contact: alice@example.com or 090-1234-5678."),
    "Contact: a####@example.com or ###-####-5678.",
    "combined mask char"
  )

  local input_text = "alice@example.com 090-1234-5678"
  local passthrough = Masker.new()
  assert_equal(passthrough:process(input_text), input_text, "masker default config")

  local email_only = Masker.new():mask_emails()
  assert_equal(email_only:process(input_text), "a****@example.com 090-1234-5678", "email only")

  local phone_only = Masker.new():mask_phones()
  assert_equal(phone_only:process(input_text), "alice@example.com ***-****-5678", "phone only")

  local both = Masker.new():mask_emails():mask_phones()
  assert_equal(both:process(input_text), "a****@example.com ***-****-5678", "both")

  local non_ascii = "連絡先: alice@example.com と 090-1234-5678"
  local expected = "連絡先: a****@example.com と ***-****-5678"
  assert_equal(both:process(non_ascii), expected, "non-ascii")
end

run()
print("All tests passed.")
